source/src/ble/cs_CrownstoneCentral.cpp

```diff
@@ -0,0 +1,663 @@
+	}
+	return retCode;
+}
+
+cs_ret_code_t CrownstoneCentral::disconnect() {
+	if (isBusy()) {
+		LOGCsCentralInfo("Busy");
+		return ERR_BUSY;
+	}
+	// No need to set current operation: BleCentral will block any other operations for us.
+	// Also, the user waits for the BleCentral event, and might get it before we get it.
+	// In which case the user wouldn't be able to connect() again, because we wouldn't have finalized the disconnect operation yet.
+	return BleCentral::getInstance().disconnect();
+}
+
+cs_ret_code_t CrownstoneCentral::write(cs_control_cmd_t commandType, uint8_t* data, uint16_t size) {
+	if (isBusy()) {
+		LOGCsCentralInfo("Busy");
+		return ERR_BUSY;
+	}
+
+	cs_ret_code_t retCode;
+
+	cs_data_t encryptedBuffer = BleCentral::getInstance().requestWriteBuffer();
+	if (encryptedBuffer.data == nullptr) {
+		return ERR_BUFFER_UNASSIGNED;
+	}
+
+	cs_data_t writeBuf = CharacteristicWriteBuffer::getInstance().getBuffer();
+	ControlPacketAccessor<> controlPacketAccessor;
+	retCode = controlPacketAccessor.assign(writeBuf.data, writeBuf.len);
+	if (retCode != ERR_SUCCESS) {
+		return retCode;
+	}
+
+	controlPacketAccessor.setProtocolVersion(CS_CONNECTION_PROTOCOL_VERSION);
+	controlPacketAccessor.setType(commandType);
+	retCode = controlPacketAccessor.setPayload(data, size);
+	if (retCode != ERR_SUCCESS) {
+		return retCode;
+	}
+	cs_data_t controlPacket = controlPacketAccessor.getSerializedBuffer();
+
+	uint16_t encryptedSize = ConnectionEncryption::getEncryptedBufferSize(controlPacket.len, ConnectionEncryptionType::CTR);
+	if (_opMode == OperationMode::OPERATION_MODE_SETUP) {
+		ConnectionEncryption::getInstance().encrypt(controlPacket, encryptedBuffer, SETUP, ConnectionEncryptionType::CTR);
+	}
+	else {
+		ConnectionEncryption::getInstance().encrypt(controlPacket, encryptedBuffer, ADMIN, ConnectionEncryptionType::CTR);
+	}
+
+	retCode = BleCentral::getInstance().write(_controlHandle, encryptedBuffer.data, encryptedSize);
+	if (retCode != ERR_WAIT_FOR_SUCCESS) {
+		return retCode;
+	}
+
+	_currentOperation = Operation::WRITE;
+	setStep(WriteControlSteps::WRITE);
+	return ERR_WAIT_FOR_SUCCESS;
+}
+
+cs_data_t CrownstoneCentral::requestWriteBuffer() {
+	if (isBusy()) {
+		return cs_data_t();
+	}
+	cs_data_t writeBuf = CharacteristicWriteBuffer::getInstance().getBuffer();
+	ControlPacketAccessor<> controlPacketAccessor;
+	cs_ret_code_t retCode = controlPacketAccessor.assign(writeBuf.data, writeBuf.len);
+	if (retCode != ERR_SUCCESS) {
+		return cs_data_t();
+	}
+	return cs_data_t(controlPacketAccessor.getPayload().data, controlPacketAccessor.getMaxPayloadSize());
+}
+
+//////////////////// Helpers ////////////////////
+
+void CrownstoneCentral::enableNotifications() {
+	cs_ret_code_t retCode = BleCentral::getInstance().writeNotificationConfig(_resultCccdHandle, true);
+	if (retCode != ERR_WAIT_FOR_SUCCESS) {
+		finalizeOperation(Operation::CONNECT, retCode);
+		return;
+	}
+	setStep(ConnectSteps::ENABLE_NOTIFICATIONS);
```

Comment:
what if the 'current step' is not 'none'?

Reply:
Hm, i was planning to enable notifications at the first write() command, and then figured out that was actually too much work for a tiny performance improvement. I guess i can move this code to the place it's being called, where the current operation and step are being checked: `onDiscoveryDone()`.